### PR TITLE
Generate default attribute value

### PR DESCRIPTION
**Pull request by postrstk**
Add generation for default attribute values.

### CODE REVIEW

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 1]
```diff
@@ -49,6 +49,43 @@ static QString pointerStorageType(const QString &typeName)
     return "QSharedPointer<" + typeName + '>';
 }
 
+static void generateDefaultAttributeValueCode(KODE::Code& result, const QString& typeName, const Converter::DefaultAttributeValue& defaultValue)
+{
+  result += "{";
+  result.indent();
+
+  if(!defaultValue.mIsBuiltin) {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: space after `if`

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 2]
```diff
@@ -355,11 +395,17 @@ QString Converter::generateMemberVariable(const QString &rawName, const QString
             KODE::Code getterCode;
             getterCode += QLatin1String("if (!") + variableName + QLatin1String("_nil)");
             getterCode.indent();
-            getterCode += QLatin1String("return &") + variableName + QLatin1Char(';');
+            defaultValue.mValue.isNull() ? getterCode += QLatin1String("return &") + variableName + QLatin1Char(';') : getterCode += QLatin1String("return ") + variableName + QLatin1Char(';');
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: Better use an `if()` here

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 3]
```diff
@@ -355,11 +395,17 @@ QString Converter::generateMemberVariable(const QString &rawName, const QString
...
             getterCode.unindent();
             getterCode += QLatin1String("else");
             getterCode.indent();
-            getterCode += "return 0;";
+
+            if(defaultValue.mValue.isNull()) {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: missing space after `if`

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 4]
```diff
@@ -355,11 +395,17 @@ QString Converter::generateMemberVariable(const QString &rawName, const QString
...
+              getterCode += "return 0;";
+            }
+            else {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: coding style: join with previous line

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 5]
```diff
@@ -370,7 +416,15 @@ QString Converter::generateMemberVariable(const QString &rawName, const QString
             getterCode.unindent();
             getterCode += QLatin1String("else");
             getterCode.indent();
-            getterCode += "return boost::optional<" + typeName + " >();";
+
+            if(defaultValue.mValue.isNull())
+            {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: coding style: join with previous line

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 6]
```diff
@@ -370,7 +416,15 @@ QString Converter::generateMemberVariable(const QString &rawName, const QString
             getterCode.unindent();
             getterCode += QLatin1String("else");
             getterCode.indent();
-            getterCode += "return boost::optional<" + typeName + " >();";
+
+            if(defaultValue.mValue.isNull())
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: if + space

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 7]
```diff
@@ -370,7 +416,15 @@ QString Converter::generateMemberVariable(const QString &rawName, const QString
...
+            {
+              getterCode += "return boost::optional<" + typeName + " >();";
+            }
+            else {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: coding style: join with previous line

File: kdwsdl2cpp/src/converter.h [hunk 8]
```diff
@@ -41,6 +41,16 @@ namespace KWSDL
 class Converter
 {
 public:
+
+  struct DefaultAttributeValue
+  {
+    DefaultAttributeValue(): mIsBuiltin(false) {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter.h]: I'd suggest to add the two arguments to the constructor, to simplify the calling code.

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 9]
```diff
@@ -310,7 +350,7 @@ QString Converter::generateMemberVariable(const QString &rawName, const QString
 
     // getter method
     QString getterTypeName = typeName;
-    if (optional) {
+    if (optional && defaultValue.mValue.isNull()) {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: I suggest to add a `defaultValue.isNull()` method to encapsulate the logic. If one day you realize that "" (isEmpty()) should also be considered as "no default value" then you'll have a single place to change.

File: kdwsdl2cpp/src/converter_complextype.cpp (hunk 1 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter_complextype.cpp]: fixed

File: kdwsdl2cpp/src/converter_complextype.cpp (hunk 2 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter_complextype.cpp]: fixed

File: kdwsdl2cpp/src/converter_complextype.cpp (hunk 3 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter_complextype.cpp]: fixed

File: kdwsdl2cpp/src/converter_complextype.cpp (hunk 4 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter_complextype.cpp]: fixed

File: kdwsdl2cpp/src/converter_complextype.cpp (hunk 5 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter_complextype.cpp]: fixed

File: kdwsdl2cpp/src/converter_complextype.cpp (hunk 6 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter_complextype.cpp]: fixed

File: kdwsdl2cpp/src/converter_complextype.cpp (hunk 7 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter_complextype.cpp]: fixed

File: kdwsdl2cpp/src/converter.h (hunk 8 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter.h]: fixed

File: kdwsdl2cpp/src/converter_complextype.cpp (hunk 9 above)
Comment (postrstk) [reply kdwsdl2cpp/src/converter_complextype.cpp]: fixed

File: kdwsdl2cpp/src/converter.h [hunk 10]
```diff
@@ -41,6 +41,21 @@ namespace KWSDL
...
+  struct DefaultAttributeValue
+  {
+    DefaultAttributeValue(bool isBuiltin = false, const QString& value = QString::null):
+        mIsBuiltin(isBuiltin),
+        mValue(value) { }
+
+    bool mIsBuiltin;
+    QString mValue;
+
+    bool isNull() {
```
[the remark this answers is not in the document]
Comment (postrstk) [reply kdwsdl2cpp/src/converter.h]: fixed

File: unittests/CMakeLists.txt [hunk 11]
```diff
@@ -91,4 +91,5 @@ add_subdirectory(onvif_org_event)
 add_subdirectory(empty_list_wsdl)
 add_subdirectory(list_restriction)
 
-add_subdirectory(kddatetime)
+add_subdirectory(default_attribute_value_wsdl)
+add_subdirectory(kddatetime)
```
[the remark this answers is not in the document]
Comment (postrstk) [reply unittests/CMakeLists.txt]: added

File: unittests/default_attribute_value_wsdl/default_attribute_value_wsdl.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      QVERIFY(actualXml.contains("errorDescription"));
+      QVERIFY(!actualXml.contains("isCancelled"));
+      QVERIFY(!actualXml.contains("intValue"));
+      QVERIFY(!actualXml.contains("doubleValue"));
+      QVERIFY(!actualXml.contains("byteArrayValue"));
+      QVERIFY(!actualXml.contains("enumerationValue"));
+    }
+
+//    Not implemented!
+//    void testDefaultDateTimeValue()
```
Comment (dfaure-kdab) [inline unittests/default_attribute_value_wsdl/default_attribute_value_wsdl.cpp]: It is implemented, actually :-)
You can remove this comment, right?
Sorry, I failed to notice that on the last review.

Comment (postrstk) [reply unittests/default_attribute_value_wsdl/default_attribute_value_wsdl.cpp]: I forgot to remove after the implementation :)
Done.

**Pull request merged**